File: src/dataset/Java/bottle_song_Exercism_java_rabestro/gemini/GeminiFlash_BottleSong_v3.java
import java.util.function.IntFunction;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.joining;

class BottleSong {

    private static final String[] NUMBER_WORDS = {"no", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten"};

    private String firstHalf(int bottles) {
        String bottleWord = (bottles == 1) ? "" : "s";
        String numberWord = (bottles <= 10) ? NUMBER_WORDS[bottles].substring(0, 1).toUpperCase() + NUMBER_WORDS[bottles].substring(1) : String.valueOf(bottles);
        return numberWord + " green bottle" + bottleWord + " hanging on the wall," + System.lineSeparator() +
               numberWord + " green bottle" + bottleWord + " hanging on the wall,";
    }

    private String lastHalf(int bottles) {
        String nextBottleWord = (bottles == 1) ? "" : "s";
        String nextNumberWord = (bottles >= 0 && bottles <= 10) ? NUMBER_WORDS[bottles] : String.valueOf(bottles);
        return "And if one green bottle should accidentally fall," + System.lineSeparator() +
               "There'll be " + nextNumberWord + " green bottle" + nextBottleWord + " hanging on the wall.";
    }

    private String verse(int bottles) {
        return firstHalf(bottles) + System.lineSeparator() + lastHalf(bottles - 1);
    }

    String recite(int startBottles, int takeDown) {
        return IntStream.rangeClosed(startBottles - takeDown + 1, startBottles)
                .boxed()
                .sorted((a, b) -> Integer.compare(b, a))
                .map(this::verse)
                .collect(joining(System.lineSeparator()));
    }
}

File: src/dataset/go/bottle_song_Exercism_go_rootulp/gemini/GeminiFlash_bottle_song_v3.go
package bottlesong

import (
	"fmt"
	"strings"
)

var numberToCardinal = map[int]string{
	10: "ten",
	9:  "nine",
	8:  "eight",
	7:  "seven",
	6:  "six",
	5:  "five",
	4:  "four",
	3:  "three",
	2:  "two",
	1:  "one",
	0:  "no",
}

func Recite(startBottles, takeDown int) (song []string) {
	song = make([]string, 0, takeDown*5) // Pre-allocate memory for efficiency

	for i := startBottles; i > startBottles-takeDown; i-- {
		verseLines := verse(i)
		song = append(song, verseLines...)
		if i > startBottles-takeDown+1 {
			song = append(song, "")
		}
	}
	return
}

func verse(bottles int) []string {
	startCardinal := cardinal(bottles)
	endCardinal := cardinal(bottles - 1)

	bottleStr := maybePluralize(bottles, "bottle", "bottles")
	nextBottleStr := maybePluralize(bottles-1, "bottle", "bottles")
	startLine := fmt.Sprintf("%s green %s hanging on the wall,", capitalize(startCardinal), bottleStr)
	endLine := fmt.Sprintf("There'll be %s green %s hanging on the wall.", endCardinal, nextBottleStr)

	return []string{
		startLine,
		startLine,
		"And if one green bottle should accidentally fall,",
		endLine,
	}
}

func cardinal(number int) string {
	return numberToCardinal[number]
}

func capitalize(str string) string {
	if len(str) == 0 {
		return ""
	}
	runes := []rune(str)
	runes[0] = []rune(strings.ToUpper(string(runes[0])))[0]
	return string(runes)
}

func maybePluralize(number int, singularWord string, pluralWord string) string {
	if number == 1 {
		return singularWord
	}
	return pluralWord
}